ndsl/dsl/dace/stree/optimizations/axis_merge.py

```diff
@@ -0,0 +1,438 @@
+
+
+def _both_same_single_axis_maps(
+    first: dst.MapScope,
+    second: dst.MapScope,
+    axis: AxisIterator,
+) -> bool:
+    return (
+        (len(first.node.params) == 1 and len(second.node.params) == 1)  # Single axis
+        and first.node.params[0] == second.node.params[0]  # Same axis
+        and _is_axis_map(first, axis)  # Correct axis
+    )
+
+
+def _can_merge_axis_maps(
+    first: dst.MapScope,
+    second: dst.MapScope,
+    axis: AxisIterator,
+) -> bool:
+    return _both_same_single_axis_maps(
+        first, second, axis
+    ) and no_data_dependencies_on_cartesian_axis(
+        first,
+        second,
+        axis,
+    )
+
+
+class InsertOvercomputationGuard(dst.ScheduleNodeTransformer):
+    def __init__(
+        self,
+        axis_as_string: str,
+        *,
+        merged_range: dace.subsets.Range,
+        original_range: dace.subsets.Range,
+    ):
+        self._axis_as_string = axis_as_string
+        self._merged_range = merged_range
+        self._original_range = original_range
+
+    def _execution_condition(self) -> CodeBlock:
+        # NOTE range.ranges are inclusive, e.g.
+        #      Range(0:4) -> ranges = (start=1, stop=3, step=1)
+        range = self._original_range
+        start = range.ranges[0][0]
+        stop = range.ranges[0][1]
+        step = range.ranges[0][2]
+        return CodeBlock(
+            f"{self._axis_as_string} >= {start} "
+            f"and {self._axis_as_string} <= {stop} "
+            f"and ({self._axis_as_string} - {start}) % {step} == 0"
+        )
+
+    def visit_MapScope(self, node: dst.MapScope) -> dst.MapScope:
+        all_children_are_maps = all(
+            [isinstance(child, dst.MapScope) for child in node.children]
+        )
+        if not all_children_are_maps:
+            if self._merged_range != self._original_range:
+                node.children = [
+                    dst.IfScope(
+                        condition=self._execution_condition(), children=node.children
+                    )
+                ]
+            return node
+
+        node.children = self.visit(node.children)
+        return node
+
+
+def _get_next_node(
+    nodes: list[dst.ScheduleTreeNode],
+    node: dst.ScheduleTreeNode,
+) -> dst.ScheduleTreeNode:
+    return nodes[list_index(nodes, node) + 1]
+
+
+def _last_node(nodes: list[dst.ScheduleTreeNode], node: dst.ScheduleTreeNode) -> bool:
+    return list_index(nodes, node) >= len(nodes) - 1
+
+
+def _sanitize_axis(axis: AxisIterator, name_to_normalize: str) -> str:
+    axis_clean = f"{axis.as_str()}"
+    pattern = f"{axis.as_str()}_[0-9]*"
+
+    return re.sub(pattern, axis_clean, name_to_normalize)
+
+
+class NormalizeAxisSymbol(dst.ScheduleNodeVisitor):
+    def __init__(self, axis: AxisIterator) -> None:
+        self.axis = axis
+
+    def visit_MapScope(
+        self,
+        map_scope: dst.MapScope,
+        axis_rpl_dict: dict[str, str] = {},
```

Comment:
```suggestion
        axis_replacements: dict[str, str] = {},
```
in my opinion, the `dict` is in the type, which leaves us almost enough characters to spell out "replacements"